File: Tyche.Android/bet/src/main/java/com/felipearpa/bet/domain/PoolGamblerBetMapper.kt
package com.felipearpa.bet.domain

import com.felipearpa.appCore.type.TeamScore

fun PoolGamblerBetResponse.toDomain() =
    PoolGamblerBet(
        poolId = this.poolId,
        gamblerId = gamblerId,
        matchId = this.matchId,
        homeTeamId = this.homeTeamId,
        homeTeamName = this.homeTeamName,
        awayTeamId = this.awayTeamId,
        awayTeamName = this.awayTeamName,
        awayTeamScore = this.awayTeamScore,
        matchScore = if (this.homeTeamScore == null || this.awayTeamScore == null) null else TeamScore(
            homeTeamValue = this.homeTeamScore,
            awayTeamValue = this.awayTeamScore
        ),
        betScore = if (this.homeTeamBet == null || this.awayTeamBet == null) null else TeamScore(
            homeTeamValue = this.homeTeamBet,
            awayTeamValue = this.awayTeamBet
        ),
        score = this.score,
        matchDateTime = this.matchDateTime
    )

File: Tyche.Backend/Felipearpa.Tyche.Pool/Felipearpa.Tyche.Pool/Domain/PoolGamblerBetMapper.fs
namespace Felipearpa.Tyche.Pool.Domain

open Felipearpa.Tyche.Pool.Type
open Felipearpa.Tyche.Pool.Data
open Felipearpa.Type

module PoolGamblerBetMapper =

    let mapToDomain (poolGamblerBetEntity: PoolGamblerBetEntity) =
        { PoolGamblerBet.PoolId = poolGamblerBetEntity.PoolId |> Ulid.newOf
          GamblerId = poolGamblerBetEntity.GamblerId |> Ulid.newOf
          MatchId = poolGamblerBetEntity.MatchId |> Ulid.newOf
          HomeTeamId = poolGamblerBetEntity.HomeTeamId |> Ulid.newOf
          HomeTeamName = poolGamblerBetEntity.HomeTeamName |> NonEmptyString100.newOf
          AwayTeamId = poolGamblerBetEntity.AwayTeamId |> Ulid.newOf
          AwayTeamName = poolGamblerBetEntity.AwayTeamName |> NonEmptyString100.newOf
          MatchScore =
            if
                poolGamblerBetEntity.HomeTeamScore.HasValue
                && poolGamblerBetEntity.AwayTeamScore.HasValue
            then
                { TeamScore.HomeTeamValue = poolGamblerBetEntity.HomeTeamScore.Value
                  AwayTeamValue = poolGamblerBetEntity.AwayTeamScore.Value }
                |> Some
            else
                None
          BetScore =
            if
                poolGamblerBetEntity.HomeTeamBet.HasValue
                && poolGamblerBetEntity.AwayTeamBet.HasValue
            then
                { TeamScore.HomeTeamValue = poolGamblerBetEntity.HomeTeamBet.Value |> BetScore.newOf
                  AwayTeamValue = poolGamblerBetEntity.AwayTeamBet.Value |> BetScore.newOf }
                |> Some
            else
                None
          Score = poolGamblerBetEntity.BetScore |> Option.ofNullable
          MatchDateTime = poolGamblerBetEntity.MatchDateTime }
